docs/examples/go-ipfs-as-a-library/main.go

```diff
@@ -0,0 +1,212 @@
+	// 	"path": "badger",
+	// }
+
+	err = fsrepo.Init(repoPath, cfg)
+	if err != nil {
+		return "", fmt.Errorf("failed to init ephemeral node: %s", err)
+	}
+
+	return repoPath, nil
+}
+
+/// ------ Spawning the node
+func createNode(ctx context.Context, repoPath string) (iCore.CoreAPI, error) {
+	// Open the repo
+	repo, err := fsrepo.Open(repoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct the node
+	node, err := core.NewNode(ctx, &core.BuildCfg{
+		Online: true,
+		// Routing: libp2p.DHTClientOption,
+		Repo: repo,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return coreapi.NewCoreAPI(node)
+}
+
+/*
+func spawnDefault(ctx context.Context) (iface.CoreAPI, error) {
+
+}
+*/
+
+/*
+func spawnEphemeral(ctx context.Context) (iface.CoreAPI, error) {
+	defaultPath, err := config.PathRoot()
+	if err != nil {
+		// shouldn't be possible
+		return nil, err
+	}
+
+	return tmpNode(ctx)
+}
+*/
+
+// Spawns a node on the default repo location, if the repo exists
+func spawnDefault(ctx context.Context) (iCore.CoreAPI, error) {
+	defaultPath, err := config.PathRoot()
+	if err != nil {
+		// shouldn't be possible
+		return nil, err
+	}
+
+	if err := setupPlugins(defaultPath); err != nil {
+		panic(err)
+	}
+
+	return createNode(ctx, defaultPath)
+}
+
+// Spawns a node to be used just for this run (i.e. creates a tmp repo)
+func spawnEphemeral(ctx context.Context) (iCore.CoreAPI, error) {
+	// Create a Temporary Repo
+	repoPath, err := createTempRepo(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create temp repo: ", err)
+	}
+
+	if err := setupPlugins(repoPath); err != nil {
+		panic(err)
+	}
+
+	// Spawning an ephemeral IPFS node
+	return createNode(ctx, repoPath)
+}
+
+// ----- Writing to disk
+func writeTo(nd files.Node, fpath string) error {
+	switch nd := nd.(type) {
+	case *files.Symlink:
+		return os.Symlink(nd.Target, fpath)
+	case files.File:
+		f, err := os.Create(fpath)
+		defer f.Close()
+		if err != nil {
+			return err
+		}
+
+		var r io.Reader = nd
+		_, err = io.Copy(f, r)
+		if err != nil {
+			return err
+		}
+		return nil
+	case files.Directory:
+		err := os.Mkdir(fpath, 0777)
+		if err != nil {
+			return err
+		}
+
+		entries := nd.Entries()
+		for entries.Next() {
+			child := filepath.Join(fpath, entries.Name())
+			if err := writeTo(entries.Node(), child); err != nil {
+				return err
+			}
+		}
+		return entries.Err()
+	default:
+		return fmt.Errorf("file type %T at %q is not supported", nd, fpath)
+	}
+}
+
+/// -------
+
+func main() {
+	fmt.Println("Getting an IPFS node running")
+
+	ctx, _ := context.WithCancel(context.Background())
+
+	/*
+		fmt.Println("Spawning node on default repo")
+		ipfs, err := spawnDefault(ctx)
+		if err != nil {
+			fmt.Println("No IPFS repo available on the default path")
+		}
+	*/
+
+	fmt.Println("Spawning node on a temporary repo")
+	ipfs, err := spawnEphemeral(ctx)
+	if err != nil {
+		panic(fmt.Errorf("failed to spawn ephemeral node: %s", err))
+	}
```

Comment:
Spawning with a temporary repo, I get the following error:

```
» go run main.go
Getting an IPFS node running
Spawning node on a temporary repo
panic: failed to spawn ephemeral node: failed to create temp repo: %!(EXTRA *errors.errorString=failed to init ephemeral node: unknown datastore type: flatfs)

goroutine 1 [running]:
main.main()
        /Users/imp/code/go-projects/src/github.com/ipfs/go-ipfs/docs/examples/go-ipfs-as-a-library/main.go:194 +0x471
exit status 2
```

Which means that the setupPlugins isn't working as expected.

Reply:
It's working as expected, but you're calling it too late. Call `setupPlugins("")` before creating the temporary repo. 

As mentioned above (https://github.com/ipfs/go-ipfs/pull/6695#discussion_r333309264) there's no need to pass in a path here.

Reply:
@aschmahmann it works on your machine because you are relying on an `ipfs init` generated repo which does all the things correctly. A new users (or a docker container) won't have the repo initiated. Either remove your ~/.ipfs or simply rename it and you will see it failing.

Reply:
nvm the above https://github.com/ipfs/go-ipfs/pull/6695#discussion_r333340414, the setupPlugins call needed to appear before the repo got created. We got it to work with https://github.com/ipfs/go-ipfs/pull/6695/commits/de61e9a6b45c9faa1416cdc259fe12a549bf3c3a